src/System.Runtime/tests/System/Double.cs

```diff
@@ -1,330 +1,402 @@
-    public static void TestParse()
+    [Theory]
+    [MemberData(nameof(ToString_TestData))]
+    public static void TestToString(double d, string format, IFormatProvider provider, string expected)
     {
-        Assert.Equal(123, double.Parse("123"));
-        Assert.Equal(-123, double.Parse("-123"));
-        //TODO: Negative tests once we get better exceptions
+        bool isDefaultProvider = (provider == null || provider == NumberFormatInfo.CurrentInfo);
+        if (string.IsNullOrEmpty(format) || format.ToUpperInvariant() == "G")
+        {
+            if (isDefaultProvider)
+            {
+                Assert.Equal(expected, d.ToString());
+                Assert.Equal(expected, d.ToString((IFormatProvider)null));
+            }
+            Assert.Equal(expected, d.ToString(provider));
+        }
+        if (isDefaultProvider)
+        {
+            Assert.Equal(expected.Replace('e', 'E'), d.ToString(format.ToUpperInvariant())); // If format is upper case, then exponents are printed in upper case
+            Assert.Equal(expected.Replace('E', 'e'), d.ToString(format.ToLowerInvariant())); // If format is lower case, then exponents are printed in upper case
+            Assert.Equal(expected.Replace('e', 'E'), d.ToString(format.ToUpperInvariant(), null));
+            Assert.Equal(expected.Replace('E', 'e'), d.ToString(format.ToLowerInvariant(), null));
+        }
+        Assert.Equal(expected.Replace('e', 'E'), d.ToString(format.ToUpperInvariant(), provider));
+        Assert.Equal(expected.Replace('E', 'e'), d.ToString(format.ToLowerInvariant(), provider));
     }
 
     [Fact]
-    public static void TestParseNumberStyle()
+    public static void TestToString_Invalid()
     {
-        Assert.Equal((double)123.1, double.Parse(string.Format("{0}", 123.1), NumberStyles.AllowDecimalPoint));
-        Assert.Equal(1000, double.Parse(string.Format("{0}", 1000), NumberStyles.AllowThousands));
-        //TODO: Negative tests once we get better exceptions
+        double d = 123;
+        Assert.Throws<FormatException>(() => d.ToString("Y")); // Invalid format
+        Assert.Throws<FormatException>(() => d.ToString("Y", null)); // Invalid format
     }
 
-    [Fact]
-    public static void TestParseFormatProvider()
+    public static IEnumerable<object[]> Parse_Valid_TestData()
     {
-        var nfi = new NumberFormatInfo();
-        Assert.Equal(123, double.Parse("123", nfi));
-        Assert.Equal(-123, double.Parse("-123", nfi));
-        //TODO: Negative tests once we get better exceptions
+        // Defaults: AllowLeadingWhite | AllowTrailingWhite | AllowLeadingSign | AllowDecimalPoint | AllowExponent | AllowThousands
+        NumberFormatInfo nullFormat = null;
+        NumberStyles defaultStyle = NumberStyles.Float;
+
+        var emptyFormat = new NumberFormatInfo();
+
+        var customFormat1 = new NumberFormatInfo();
+        customFormat1.CurrencySymbol = "$";
+        customFormat1.CurrencyGroupSeparator = ",";
```

Comment:
@hughbe, I see you pushed an "Address PR feedback" commit after I made this comment, but it doesn't appear to have been addressed.  Was that on purpose?